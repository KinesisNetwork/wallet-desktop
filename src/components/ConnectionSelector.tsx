import * as React from 'react'

import { Connection } from '@types'

export interface Props {
  connections: Connection[]
  activeConnection: number
  selectConnection: (conn: Connection) => any
}

export class ConnectionSelector extends React.Component<Props> {
  displayName = 'ConnectionSelector'

  constructor(props) {
    super(props)
  }

  renderConnections = () => {
    return this.props.connections.map((conn, i) => {
      const isActive = this.props.activeConnection === i
      return (
        <ConnectionButton
          key={i}
          connection={conn}
          isActive={isActive}
          selectConnection={this.props.selectConnection}
        />
      )
    })
  }

  render() {
    return (
      <section className='section'>
        <div>
          <span className='icon is-large'>
            <i className='fas fa-wifi' />
          </span>
          <h1 className='title is-4 heading primary-font'>Select a Network</h1>
        </div>
        {this.renderConnections()}
      </section>
    )
  }
}

interface ConnectionButtonProps {
  connection: Connection
  isActive: boolean
  selectConnection: (conn: Connection) => any
}

class ConnectionButton extends React.Component<ConnectionButtonProps> {
  displayName = 'ConnectionButton'

  constructor(props) {
    super(props)
  }

  handleClick = (ev) => {
    ev.preventDefault()
    this.props.selectConnection(this.props.connection)
  }

  render() {
    const { connection, isActive } = this.props
    return (
      <a className='level' onClick={this.handleClick}>
        <div className='level-left has-text-left'>
          <div className='content'>
            <strong>{connection.name}</strong>
            <p>
              {connection.horizonURL}<br />
              {connection.networkPassphrase}
            </p>
          </div>
        </div>
        <div className='level-right'>
          {isActive && <span className='icon has-text-success'><i className='fas fa-check-circle' /></span>}
        </div>
      </a>
    )
  }
}
