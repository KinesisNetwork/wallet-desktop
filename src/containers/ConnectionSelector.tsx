import * as React from 'react'
import { connect } from 'react-redux'

import { EditableText } from '@components/EditableText'

import {
  handleConnectionFormChange,
  selectConnectedCurrency,
  selectConnectedStage,
  selectForEditConnection,
  selectUpdatingCurrency,
  stopEditingConnection,
} from '@actions'
import { getCurrentConnectionForEditing } from '@selectors'
import { RootState } from '@store'
import { ConnectionStage, Currency } from '@types'

export const mapStateToProps = ({ connections }: RootState) => ({
  isEditing: connections.updating.isEditing,
  connectionToEdit: getCurrentConnectionForEditing(connections),
  currentCurrency: connections.updating.selectedCurrency,
  currentStage: connections.currentStage,
})

const mapDispatchToProps = {
  selectConnectedStage,
  selectConnectedCurrency,
  selectForEditConnection,
  handleConnectionFormChange,
  stopEditingConnection,
  selectUpdatingCurrency,
}

type Props = typeof mapDispatchToProps & ReturnType<typeof mapStateToProps>

const ConnectionSelectorComponent: React.SFC<Props> = props => (
  <nav className="panel">
    <div className="panel-heading">
      <div className="field is-grouped is-horizontal">
        <div className="field-label is-normal">
          <label>Network</label>
        </div>
        <div className="field-body">
          <div className="field has-addons has-addons-right">
            <div className="control">
              <button
                className={[
                  'button',
                  ...(props.currentStage === ConnectionStage.mainnet
                    ? ['is-active', 'is-success']
                    : []),
                ].join(' ')}
                onClick={() => props.selectConnectedStage(ConnectionStage.mainnet)}
              >
                Main
              </button>
            </div>
            <div className="control">
              <button
                className={[
                  'button',
                  ...(props.currentStage === ConnectionStage.testnet
                    ? ['is-active', 'is-success']
                    : []),
                ].join(' ')}
                onClick={() => props.selectConnectedStage(ConnectionStage.testnet)}
              >
                Test
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div className="panel-tabs">
      <a
        className={props.currentCurrency === Currency.KAU ? 'is-active' : ''}
        onClick={() => props.selectUpdatingCurrency(Currency.KAU)}
      >
        {Currency.KAU}
      </a>
      <a
        className={props.currentCurrency === Currency.KAG ? 'is-active' : ''}
        onClick={() => props.selectUpdatingCurrency(Currency.KAG)}
      >
        {Currency.KAG}
      </a>
      <a
        className={props.currentCurrency === Currency.KEM ? 'is-active' : ''}
        onClick={() => props.selectUpdatingCurrency(Currency.KEM)}
      >
        {Currency.KEM}
      </a>
    </div>
    <div className="panel-block is-block">
      <div className="field is-horizontal">
        <div className="field-label is-normal">
          <label className="label">Endpoint</label>
        </div>
        <div className="field-body">
          <div className="field">
            <EditableText
              isEditing={props.isEditing === 'endpoint'}
              value={props.connectionToEdit.endpoint}
              onChangeHandler={ev =>
                props.handleConnectionFormChange({
                  field: 'endpoint',
                  newValue: ev.currentTarget.value,
                  currentCurrency: props.currentCurrency,
                  currentStage: props.currentStage,
                })
              }
              onStartEditing={() => props.selectForEditConnection('endpoint')}
              onStopEditing={() => props.stopEditingConnection()}
            />
          </div>
        </div>
      </div>
    </div>
    <div className="panel-block is-block">
      <div className="field is-horizontal">
        <div className="field-label is-normal">
          <label className="label">Passphrase</label>
        </div>
        <div className="field-body">
          <div className="field">
            <EditableText
              isEditing={props.isEditing === 'passphrase'}
              value={props.connectionToEdit.passphrase}
              onChangeHandler={ev =>
                props.handleConnectionFormChange({
                  field: 'passphrase',
                  newValue: ev.currentTarget.value,
                  currentCurrency: props.currentCurrency,
                  currentStage: props.currentStage,
                })
              }
              onStartEditing={() => props.selectForEditConnection('passphrase')}
              onStopEditing={() => props.stopEditingConnection()}
            />
          </div>
        </div>
      </div>
    </div>
  </nav>
)

export const ConnectionSelector = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ConnectionSelectorComponent)
