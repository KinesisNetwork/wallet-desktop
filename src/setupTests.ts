import { configure } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'

export default configure({ adapter: new Adapter() })
