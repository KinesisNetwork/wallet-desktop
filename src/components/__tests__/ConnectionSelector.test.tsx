import { shallow } from 'enzyme';
import * as React from 'react'

import { ConnectionSelector } from '@components/ConnectionSelector'
import '../../setupTests'

describe('ConnectionSelector', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <ConnectionSelector
        connections={[
          {
            name: 'Kinesis KAU Testnet',
            networkPassphrase: 'qwe',
            horizonURL: 'https://asd.io'
          }
        ]}
        activeConnection={0}
        selectConnection={() => null}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('renders a primary heading and an icon', () => {
    const wrapper = shallow(
      <ConnectionSelector
        connections={[
          {
            name: 'Kinesis KAU Testnet',
            networkPassphrase: 'qwe',
            horizonURL: 'https://asd.io'
          }
        ]}
        activeConnection={0}
        selectConnection={() => null}
      />
    )

    expect(wrapper.find('h1')).toHaveLength(1)
    expect(wrapper.find('h1').text()).toEqual('Select a Network')
    expect(wrapper.find('.icon')).toHaveLength(1)
    expect(wrapper.find('.icon').children().hasClass('fa-wifi')).toBe(true)
  })

  it('should two ConnectionButton components if two connections are fed', () => {
    const wrapper = shallow(
      <ConnectionSelector
        connections={[
          {
            name: 'Kinesis KAU Testnet',
            networkPassphrase: 'qwe',
            horizonURL: 'https://asd.io'
          },
          {
            name: 'Kinesis KAG Testnet',
            networkPassphrase: 'rtr',
            horizonURL: 'https://zxc.io'
          }
        ]}
        activeConnection={0}
        selectConnection={() => null}
      />
    )

    expect(wrapper.children()).toHaveLength(3)
    expect(wrapper.find('ConnectionButton')).toHaveLength(2)
  })
})
