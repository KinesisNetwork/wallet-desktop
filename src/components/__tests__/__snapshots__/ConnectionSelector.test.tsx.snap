// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectionSelector renders correctly 1`] = `
<section
  className="section"
>
  <div>
    <span
      className="icon is-large"
    >
      <i
        className="fas fa-wifi"
      />
    </span>
    <h1
      className="title is-4 heading primary-font"
    >
      Select a Network
    </h1>
  </div>
  <ConnectionButton
    connection={
      Object {
        "horizonURL": "https://asd.io",
        "name": "Kinesis KAU Testnet",
        "networkPassphrase": "qwe",
      }
    }
    isActive={true}
    key="0"
    selectConnection={[Function]}
  />
</section>
`;
